### DISCUSSION TITLE
Pre-fill form fields & create missing items

### DISCUSSION THREAD

**Discussion by SeanDylanGoff** [directus/directus#13939] [Feature Requests]
We have a use-case where we control Directus using a separate webpage.
We can already link directly to a specific item, and we can also change the route without reloading the page using a simple extension that issues $router.push's when told so over postMessage.

What we're missing is the ability to pre-populate fields (for example with a query param like ?data={"field1": "abc", "field2": 5})
For this project we use the "manual string" PK option. We'd also like to be able to link to /admin/content/ourdata/abc, and for Directus to automatically open /admin/content/ourdata/+?data={"id": "abc"} if "abc" does not yet exist.

While quite easy to implement in the core - this is impossible with extensions only.
Would the maintainers be open to a PR on this?

**Comment by rijkvanzanten** [comment]
The ability to pre-fill fields would be pretty useful to have! I am a little worried about the max-length of URLs becoming a problem though 🤔 JSON is pretty verbose (with all the extra quotes and whatnot. Maybe we should go with a more "traditional" query syntax like `?data[field]=value` ?

**Reply by SeanDylanGoff** [reply]
True, json was about 50% bigger in some simple test i did. And i think there's nothing that can't be represented in QS - and it's consistent with how the API handles params.

What do you think about the second idea, the redirecting on non-existing key? Would a PR be welcome with those two features?

**Reply by rijkvanzanten** [reply]
> What do you think about the second idea, the redirecting on non-existing key?

I'm a little hesitant on this, as it's way more opinionated and removes the (valid) use case of having a 404 for unknown items. It would also only work for items with manual primary keys, as it would break for auto-int / uuid use cases 🤔

**Reply by SeanDylanGoff** [reply]
Understandable, i forgot to mention that we put it behind a query param in our implementation. So something like ?createIfNotExists=true 

Also, shortly after writing my last reply i realised that query string encoding does have some problems:
1. JSON field contents would still have to be JSON to not lose typing. This makes serializing very weird and manual.
2. To set a field explicitly to NULL we'd probably have to use sth like val=%00, which again makes serializing even weirder.
3. Encoding an empty array (to truncate a O2M/M2M) seems impossible without some ugly constants

So maybe it's better to just use JSON anyways? Query param length will always be a limiting factor anyways (though they can be surprisingly long), so maybe it's an OK tradeoff?

**Reply by rijkvanzanten** [reply]
Lets start with just the JSON pre-fill, as I'm not too sure on the 80/20 of the create-when-doesnt-exist query 🤔 

Regarding query param format for the data, going with JSON would make sense, but is very "ugly" and not very human readable. It would be nice if we could support both `data={"json": "x"}` and `data[json]=x`, though the type casting might become tricky like you mentioned

**Reply by SeanDylanGoff** [reply]
What can i do to get this accepted/converted into an issue? The related PR was rejected for missing an issue (i just linked the discussion)

**Reply by rijkvanzanten** [reply]
> What can i do to get this accepted/converted into an issue? The related PR was rejected for missing an issue (i just linked the discussion)

So far, this feature request hasn't really seen enough engagement to warrant the addition to and long term maintenance in core unfortunately, so I don't think we're looking to add it as a feature at this point..

**Comment by azrikahar** [comment]
Linking related discussion for discoverability: #10232

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
